benchmark/components/fnt.cpp

```diff
@@ -0,0 +1,749 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <chrono>
+#include <fstream>
+#include <functional>
+#include <iostream>
+#include <vector>
+
+#include <benchmark/benchmark.h>
+
+#include "arith.h"
+#include "core.h"
+#include "fec_rs_fnt.h"
+#include "fft_2n.h"
+#include "gf_prime.h"
+#include "misc.h"
+#include "simd/simd.h"
+#include "vec_buffers.h"
+
+namespace arith = quadiron::arith;
+namespace vec = quadiron::vec;
+namespace gf = quadiron::gf;
+namespace fft = quadiron::fft;
+namespace fec = quadiron::fec;
+namespace simd = quadiron::simd;
+
+// Vector length: number of vectors (registers) per buffer
+const size_t MIN_VEC_LEN = 8;
+const size_t MAX_VEC_LEN = 512;
+// Block size
+const size_t MIN_BLOCK_SIZE = 32 * MAX_VEC_LEN;
+const size_t MAX_BLOCK_SIZE = 128 * MAX_VEC_LEN;
+// File size
+const size_t FILE_SIZE_BYTES = 51200;
+// Code length
+const size_t MIN_CODE_LEN = 16;
+const size_t MAX_CODE_LEN = 1024;
+// "Inversed" code rate
+// k = {1, 2, .., MAX_INV_RATE-1} * n / MAX_INV_RATE
+const unsigned MAX_INV_RATE = 4;
+
+template <typename T>
+class PerfFnt : public benchmark::Fixture {
+  public:
+    T q;
+    unsigned word_size;
+    std::unique_ptr<std::uniform_int_distribution<uint32_t>> distribution;
+
+    size_t n;
+    size_t size;
+
+    std::unique_ptr<vec::Buffers<T>> wk_data = nullptr;
+    std::unique_ptr<gf::Field<T>> gf = nullptr;
+    std::vector<std::string> d_files;
+    std::vector<std::string> c_files;
+    std::vector<std::string> r_files;
+
+    PerfFnt()
+    {
+        if (sizeof(T) == 2) {
+            this->q = 257;
+            word_size = 1;
+        } else if (sizeof(T) == 4) {
+            this->q = static_cast<T>(65537);
+            word_size = 2;
+        } else {
+            throw "Wrong TypeParam for PerfFnt tests";
+        }
+
+        this->distribution =
+            std::make_unique<std::uniform_int_distribution<uint32_t>>(0, q - 1);
+
+        gf = gf::alloc<gf::Field<T>, gf::Prime<T>>(q);
+
+        n = 2;
+        size = simd::countof<T>() * MAX_VEC_LEN;
+        wk_data = std::make_unique<vec::Buffers<T>>(n, size);
+        randomize_data(*wk_data);
+
+        prepare_codec_streams();
+    }
+
```

Comment:
This seems duplicated from `benchmark/components/simd_fnt.cpp `, we should factor it out into an helper

Reply:
yes, it duplicates :)

Reply:
I refactor these codes into classes. It'll be easier to add more test later.